central/declarativeconfig/updater/auth_provider_updater.go

```diff
@@ -39,3 +55,33 @@ func (u *authProviderUpdater) Upsert(ctx context.Context, m proto.Message) error
 	}
 	return nil
 }
+
+func (u *authProviderUpdater) DeleteResources(ctx context.Context, resourceIDsToSkip ...string) ([]string, error) {
+	resourcesToSkip := set.NewFrozenStringSet(resourceIDsToSkip...)
+
+	authProviders, err := u.authProviderDS.GetAuthProvidersFiltered(ctx, func(authProvider *storage.AuthProvider) bool {
+		return authProvider.GetTraits().GetOrigin() == storage.Traits_DECLARATIVE &&
+			!resourcesToSkip.Contains(authProvider.GetId())
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving declarative auth providers")
+	}
+
+	var authProviderDeletionErr *multierror.Error
+	var authProviderIDs []string
+	for _, authProvider := range authProviders {
+		if err := u.authProviderRegistry.DeleteProvider(ctx, authProvider.GetId(), true, true); err != nil {
+			authProviderDeletionErr = multierror.Append(authProviderDeletionErr, err)
+			authProviderIDs = append(authProviderIDs, authProvider.GetId())
+
+			u.reporter.UpdateIntegrationHealthAsync(utils.IntegrationHealthForProtoMessage(authProvider, "", err,
+				u.idExtractor, u.nameExtractor))
+			continue
+		}
+		// TODO(ROX-14700): This currently deletes also declarative groups and should resolve these references.
+		if err := u.groupDS.RemoveAllWithAuthProviderID(ctx, authProvider.GetId(), true); err != nil {
```

Comment:
Yeah, that's true for these "transient" kind of things.

But, the alternative would be that we never delete those groups, and since the behavior when deleting an auth provider created via API is to also delete all associated groups, why would we suddenly change the behavior?

One major issue with the groups is that they aren't really discoverable for users. They don't have a separate page within UI, and discovering them via API also isn't as easy (one would have to know the auth provider ID, however since we are in "declarative config workflow", you typically only have the name identifier). So, how would a customer _ever_ get to delete those in an acceptable UX flow?

I'd propose to keep deleting the imperative groups, and specifically mention them as such within the documentation. If they really want to keep those kind of things, why not make this declarative groups to begin with?
(I know there will be the ACSCS use-case, but that doesn't warrant for me personally to change the behavior of deleting auth providers).

Reply:
I see two ways of dealing with such groups:
1. Delete imperative groups after a certain time/number of reconciliation attempts they become orphaned. That would not completely solve the problem, but definitely reduce the probability of such a problem - as you mentioned, it only applies to "transient" behaviour.
2. Create a button in `System Health` that cleans such groups explicitly. That is more complex approach, but at the same time can be extended to other orphaned resources - imperative roles for which we try to delete declarative permission set, for example.

We can keep it as it is now, and resolve it within ROX-14700.